web-api/src/app/state.rs

```diff
@@ -27,25 +42,62 @@ where
     pub(crate) config: A::Config,
     #[deref]
     pub(crate) extension: A::Extension,
-    pub(crate) storage: A::Storage,
+    storage_builder: Arc<StorageBuilder>,
 }
 
 impl<A> AppState<A>
 where
     A: Application,
 {
+    pub(super) fn attach_to<T>(self: Arc<Self>, mut app: App<T>) -> App<T>
+    where
+        T: ServiceFactory<ServiceRequest, Config = (), Error = actix_web::Error, InitError = ()>,
+    {
+        app = app.app_data(self.storage_builder.clone());
+        app = app.app_data(Data::from(self));
+        app = app.configure(A::configure_service);
+        app
+    }
+
     pub(super) async fn create(config: A::Config) -> Result<Self, SetupError> {
         let extension = A::create_extension(&config)?;
-        let storage = A::setup_storage(config.as_ref()).await?;
-
+        let enable_legacy_tenant =
+            identity::<&tenants::Config>(config.as_ref()).enable_legacy_tenant;
+        let storage_builder =
+            Arc::new(Storage::builder(config.as_ref(), enable_legacy_tenant).await?);
         Ok(Self {
             config,
             extension,
-            storage,
+            storage_builder,
         })
     }
 
     pub(super) async fn close(self: Arc<Self>) {
-        A::close_storage(&self.storage).await;
+        self.storage_builder.close().await;
     }
 }
+
+/// Extract tenant specif state.
+///
+/// For now this only extracts storage.
+pub(crate) struct TenantState(pub(crate) Storage);
+
+impl FromRequest for TenantState {
+    type Error = Error;
+
+    type Future = Ready<Result<Self, Self::Error>>;
+
+    fn from_request(request: &HttpRequest, _payload: &mut Payload) -> Self::Future {
+        ready(extract_tenant_state(request))
+    }
+}
+
+fn extract_tenant_state(request: &HttpRequest) -> Result<TenantState, Error> {
+    let tenant_id = RequestContext::try_extract_from_request(request, |ctx| ctx.tenant_id)
+        .map_err(InternalError::from_std)?;
+    let storage = request
+        .app_data::<Arc<StorageBuilder>>()
+        .ok_or_else(|| InternalError::from_message("Arc<StorageBuilder> missing"))?
+        .build_for(tenant_id)?;
+    Ok(TenantState(storage))
```

Comment:
It's a sub component of the `FromRequest` implementation for `TenantId` not using self which wouldn't exist if `catch {}` blocks where already a thing. I don't think it should be attached to `TenantId` it also doesn't bring any benefit to attach it there IMHO.

And I still would use `Ok(Self(storage))` instead of `.map(Self)` as it better coveys the meaning of the code (creating a `TenantState` instance from all of it's components, which currently happens to only be storage.